### PR TITLE
Add `miri_start` support

### LINKED ISSUE #3758: Support miri-specific entry point / start function for no-std binaries
For no-std binaries, Miri currently relies on the `start` feature, but that feature [doesn't really make a lot of sense](https://github.com/rust-lang/rust/issues/29633#issuecomment-2088596042) and should likely be removed. It also causes [confusion](https://github.com/rust-lang/miri/issues/3498) for Miri users.

So we should instead say that in a no-std binary, we Miri will start the program by jumping to a function with link_name `miri_start`, and then if someone has a no-std binary they want to run in Miri, they can add something like
```rust
#[cfg(miri)]
#[no_mangle]
fn miri_start() {
  // Call the actual start function that your project implements, based on your target's conventions.
}
```

### PR DESCRIPTION
**Pull request by primoly**
This PR uses a function with the exported symbol `miri_start` as a drop-in alternative to `#[start]`. So the signature stays the same as suggested in [this comment](https://github.com/rust-lang/miri/issues/3498#issuecomment-2088560526). <del>I’ve also removed Miri’s restriction  to only work on bin crates as I don’t think this is necessary anymore.</del>

Closes #3758

### CODE REVIEW

File: src/bin/miri.rs [hunk 1]
```diff
@@ -78,14 +78,22 @@ impl rustc_driver::Callbacks for MiriCompilerCalls {
 
             let early_dcx = EarlyDiagCtxt::new(tcx.sess.opts.error_format);
             init_late_loggers(&early_dcx, tcx);
-            if !tcx.crate_types().contains(&CrateType::Executable) {
```
Comment (Noratrieb) [inline src/bin/miri.rs:81]: I think this restriction should stay, it doesn't make a lot of sense to execute a library.

File: src/bin/miri.rs [hunk 2]
```diff
@@ -78,14 +78,22 @@ impl rustc_driver::Callbacks for MiriCompilerCalls {
...
+                    if sym.symbol_name_for_local_instance(tcx).name == "miri_start" {
+                        Some(sym)
+                    } else {
+                        None
+                    }
+                });
+                if let Some(ExportedSymbol::NonGeneric(id)) = sym {
+                    (*id, EntryFnType::Start)
+                } else {
+                    tcx.dcx().fatal("miri can only run programs that have a main function or one exported with symbol miri_start");
```
Comment (Noratrieb) [inline src/bin/miri.rs:95]: the error message should probably differentiate between no symbol found and a generic symbol found

Comment (Noratrieb) [reply src/bin/miri.rs:95]: I'd then reword the not found one to
> Miri can only run programs that have a main function or export the symbol `miri_start` (for example, with `#[no_mangle] fn miri_start()`)"

File: src/bin/miri.rs [hunk 3]
```diff
@@ -85,7 +90,55 @@ impl rustc_driver::Callbacks for MiriCompilerCalls {
             let (entry_def_id, entry_type) = if let Some(entry_def) = tcx.entry_fn(()) {
                 entry_def
             } else {
-                tcx.dcx().fatal("miri can only run programs that have a main function");
+                let sym = tcx.exported_symbols(LOCAL_CRATE).iter().find_map(|(sym, _)| {
```
Comment (RalfJung) [inline src/bin/miri.rs:93]: Please factor this new entry point logic into a separate function.

File: tests/fail/no_main.rs [hunk 4]
```diff
@@ -1,2 +1,9 @@
-//@error-in-other-file: miri can only run programs that have a main function
+//@error-in-other-file: Miri can only run programs that have a main function.
+//Alternatively, you can export a `miri_start` function:
+//
+//#[cfg(miri)]
+//#[no_mangle]
+//fn miri_start(argc: isize, argv: *const *const u8) -> isize {
+//    // Call the actual start function that your project implements, based on your target's conventions.
+//}
```
Comment (RalfJung) [inline tests/fail/no_main.rs:8]: This should not be needed, `error-in-other-file` is not multi-line.

**Review by RalfJung — COMMENTED** [review]
This is very nice, thanks. :) I just have some minor comments.

File: src/bin/miri.rs [hunk 5]
```diff
@@ -404,6 +354,61 @@ fn jemalloc_magic() {
     }
 }
 
+fn entry_fn(tcx: TyCtxt<'_>) -> (DefId, EntryFnType) {
+    if let Some(entry_def) = tcx.entry_fn(()) {
+        entry_def
+    } else {
+        let sym = tcx.exported_symbols(LOCAL_CRATE).iter().find_map(|(sym, _)| {
```
Comment (RalfJung) [inline src/bin/miri.rs:361]:
```suggestion
        // Look for a symbol in the local crate named `miri_start`, and threat that as the entry point.
        let sym = tcx.exported_symbols(LOCAL_CRATE).iter().find_map(|(sym, _)| {
```

File: src/bin/miri.rs [hunk 6]
```diff
@@ -404,6 +354,61 @@ fn jemalloc_magic() {
...
+                    (*id, EntryFnType::Start)
+                } else {
+                    tcx.dcx().fatal(
+                        "`miri_start` must have the following signature:\
+                        \nfn miri_start(argc: isize, argv: *const *const u8) -> isize",
+                    );
+                }
+            }
+            _ => {
+                tcx.dcx().fatal("Miri can only run programs that have a main function.\
```
Comment (RalfJung) [inline src/bin/miri.rs:399]: Please put the `\n` at the ends of these lines, rather than the beginning of the next lines.

**Comment by RalfJung** [comment]
Please also squash this into a single commit. :)

**Comment by RalfJung** [comment]
Thanks. :)

@bors r+

**Pull request merged**